[Inductor] [Quant] Re-structure Quantization testcase pattern matcher check

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112570

**Summary**
This Diff re-structures Quantization testcase pattern matcher check. Instead of checking all the pattern matched in the Inductor, we will only check the core pattern match count and node numbers such as: dequant promotion, QConv/Linear Unary and QConv Binary.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_q
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

leslie-fang-intel committed:
[Inductor] [Quant] Re-structure Quantization testcase pattern matcher check

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [Quant] Re-structure Quantization testcase pattern matcher check"


**Summary**
This Diff re-structures Quantization testcase pattern matcher check. Instead of checking all the pattern matched in the Inductor, we will only check the core pattern match count and node numbers such as: dequant promotion, QConv/Linear Unary and QConv Binary.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_q
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [Quant] Re-structure Quantization testcase pattern matcher check"


**Summary**
This Diff re-structures Quantization testcase pattern matcher check. Instead of checking all the pattern matched in the Inductor, we will only check the core pattern match count and node numbers such as: dequant promotion, QConv/Linear Unary and QConv Binary.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_q
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Inductor] [Quant] Re-structure Quantization testcase pattern matcher check"


**Summary**
This Diff re-structures Quantization testcase pattern matcher check. Instead of checking all the pattern matched in the Inductor, we will only check the core pattern match count and node numbers such as: dequant promotion, QConv/Linear Unary and QConv Binary.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_q
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jerryzh168 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -834,6 +843,10 @@ def qconv_weight_prepack(match: Match, *args, **kwargs):
             if clone_node is not None:
                 graph.erase_node(clone_node)
             graph.erase_node(dequant_per_channel)
+            counters["inductor"]["qconv2d_weight_prepack_matcher_count"] += 1
```
is it OK to use this? I have a different util for checking nodes actually, would that work: https://github.com/pytorch/pytorch/blob/main/test/quantization/pt2e/test_quantize_pt2e.py#L394

leslie-fang-intel replied:
Hi @jerryzh168, thanks for the comments. 

- I think add and check the `counters` is more like the way other inductor UT did as:   https://github.com/pytorch/pytorch/blob/ff35e1e45b9869fb3739469da96ea016a4ab739a/test/inductor/test_split_cat_fx_passes.py#L88-L91 and https://github.com/pytorch/pytorch/blob/8191fb3e060f600090c839d6b86c28dbb687b679/torch/_inductor/fx_passes/split_cat.py#L99

- Per my understanding, I am afraid this function (https://github.com/pytorch/pytorch/blob/main/test/quantization/pt2e/test_quantize_pt2e.py#L394) may not work in this case, since we may not able to get the model after fusion passes but before lowering into CPP Backend.

jerryzh168 replied:
OK sounds good, thanks for the explanation

leslie-fang-intel commented:
Hi @jerryzh168 @jgong5, could you help to take a look of this PR again for simplifying the match checker in `test_mkldnn_pattern_matcher`? If we can land this PR firstly, I think it can also help us simplifying the changes made in several PRs (like https://github.com/pytorch/pytorch/pull/112550) for enabling of int8-mixed-bf16 quantization.

jgong5 commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -1030,29 +1045,29 @@ def forward(self, x):
+                counters["inductor"]["qlinear_weight_prepack_matcher_count"], 3
+            )
+            self.assertEqual(
+                counters["inductor"]["qlinear_weight_prepack_matcher_nodes"], 18
```
Do you think it's possible to have a function to calculate the number of matcher nodes from the pattern? Then, we don't have to hard-code the numbers here.

leslie-fang-intel replied:
Yean, I think it depends on how to write a graph analysis function with specific heuristic for calculating the expected pattern kind and pattern number given a input fx graph (after quantization flow).

If we we can get that pattern kind and numbers, then we can define a pre-known table match from pattern kind to pattern matcher node number, for example
```
{conv2d_weight_prepack: 6, conv2d_relu_fp32_output: 2, conv2d_relu_int8_output: 7, .....}
```

jgong5 reviewed (approved):
LGTM. One minor suggestion FYI.

leslie-fang-intel committed:
Update on "[Inductor] [Quant] Re-structure Quantization testcase pattern matcher check"


**Summary**
This Diff re-structures Quantization testcase pattern matcher check. Instead of checking all the pattern matched in the Inductor, we will only check the core pattern match count and node numbers such as: dequant promotion, QConv/Linear Unary and QConv Binary.

**TestPlan**
```
python -m pytest test_mkldnn_pattern_matcher.py -k test_q
```

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jerryzh168 commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -447,18 +455,21 @@ def forward(self, x):
+                counters["inductor"]["qconv2d_weight_prepack_matcher_count"], 1
+            )
+            self.assertEqual(
+                counters["inductor"]["qconv2d_weight_prepack_matcher_nodes"], 6
```
do you need this?

leslie-fang-intel replied:
Ideally, we only need to check `matcher_count`. `matcher_nodes` is a stricter restriction to check a node number in each matcher pattern.

jerryzh168 commented on test/inductor/test_mkldnn_pattern_matcher.py:
```diff
@@ -447,18 +455,21 @@ def forward(self, x):
-        #    [convert_element_type_1, sub, mul_1, dequantize_per_channel, clone, convolution]
+        def matcher_check_fn():
+            # 1. Dequant-Conv2D pattern matched in QConv2D weight prepack * 1
+            #    [convert_element_type_1, sub, mul_1, dequantize_per_channel, clone, convolution]
```
how do we get this list?

leslie-fang-intel replied:
By printing the `m.nodes` here, we can the nodes list in each matched pattern https://github.com/pytorch/pytorch/blob/042445b7d3c7e7d4c1c66bcb3e5d35498bd2a4fb/torch/_inductor/pattern_matcher.py#L1137

jerryzh168 reviewed (approved):
not sure why you need the matcher_nodes count

leslie-fang-intel commented:
> not sure why you need the matcher_nodes count

Thanks. Yean, `matcher_nodes` is not a must have. We can only keep the `matcher_nodes` checker for some important pattern.

leslie-fang-intel commented:
@pytorchbot merge

leslie-fang-intel commented:
@pytorchbot merge
